test/unittests/floating_point_utils_test.cpp

```diff
@@ -0,0 +1,179 @@
+    EXPECT_EQ(FP(FP64::Limits::quiet_NaN()).as_uint(), 0x7FF'8000000000000);
+    EXPECT_EQ(FP(-FP64::Limits::quiet_NaN()).as_uint(), 0xFFF'8000000000000);
+}
+
+TEST(floating_point_utils, float_as_uint)
+{
+    EXPECT_EQ(FP(0.0f).as_uint(), 0x00000000);
+    EXPECT_EQ(FP(-0.0f).as_uint(), 0x80000000);
+    EXPECT_EQ(FP(FP32::Limits::infinity()).as_uint(), 0x7F800000);
+    EXPECT_EQ(FP(-FP32::Limits::infinity()).as_uint(), 0xFF800000);
+    EXPECT_EQ(FP(FP32::Limits::max()).as_uint(), 0x7F7FFFFF);
+    EXPECT_EQ(FP(-FP32::Limits::max()).as_uint(), 0xFF7FFFFF);
+    EXPECT_EQ(FP(FP32::Limits::min()).as_uint(), 0x00800000);
+    EXPECT_EQ(FP(-FP32::Limits::min()).as_uint(), 0x80800000);
+    EXPECT_EQ(FP(FP32::Limits::denorm_min()).as_uint(), 0x00000001);
+    EXPECT_EQ(FP(-FP32::Limits::denorm_min()).as_uint(), 0x80000001);
+    EXPECT_EQ(FP(1.0f).as_uint(), 0x3F800000);
+    EXPECT_EQ(FP(-1.0f).as_uint(), 0xBF800000);
+    EXPECT_EQ(FP(std::nextafter(1.0f, 0.0f)).as_uint(), 0x3F7FFFFF);
+    EXPECT_EQ(FP(std::nextafter(-1.0f, 0.0f)).as_uint(), 0xBF7FFFFF);
+    EXPECT_EQ(FP(FP32::Limits::quiet_NaN()).as_uint(), 0x7FC00000);
+    EXPECT_EQ(FP(-FP32::Limits::quiet_NaN()).as_uint(), 0xFFC00000);
+}
+
+TEST(floating_point_utils, double_from_uint)
+{
+    EXPECT_EQ(FP(uint64_t{0x0000000000000000}).value, 0.0);
+    EXPECT_EQ(FP(uint64_t{0x8000000000000000}).value, -0.0);
+    EXPECT_EQ(FP(uint64_t{0x3FF'000000000DEAD}).value, 0x1.000000000DEADp0);
+    EXPECT_EQ(FP(uint64_t{0xBFF'000000000DEAD}).value, -0x1.000000000DEADp0);
+    EXPECT_EQ(FP(uint64_t{0x7FF'0000000000000}).value, FP64::Limits::infinity());
+    EXPECT_EQ(FP(uint64_t{0xFFF'0000000000000}).value, -FP64::Limits::infinity());
+}
+
+TEST(floating_point_utils, float_from_uint)
+{
+    EXPECT_EQ(FP(uint32_t{0x00000000}).value, 0.0f);
+    EXPECT_EQ(FP(uint32_t{0x80000000}).value, -0.0f);
+    EXPECT_EQ(FP(uint32_t{0x3FEF5680}).value, 0x1.DEADp0f);
+    EXPECT_EQ(FP(uint32_t{0xBFEF5680}).value, -0x1.DEADp0f);
+    EXPECT_EQ(FP(uint32_t{0x7F800000}).value, FP32::Limits::infinity());
+    EXPECT_EQ(FP(uint32_t{0xFF800000}).value, -FP32::Limits::infinity());
+}
+
+TEST(floating_point_utils, double_nan_payload)
+{
+    constexpr auto inf = FP64::Limits::infinity();
+    constexpr auto qnan = FP64::Limits::quiet_NaN();
+
+    EXPECT_EQ(FP(0.0).nan_payload(), 0);
+    EXPECT_EQ(FP(FP64::Limits::signaling_NaN()).nan_payload(), 0x4000000000000);
+    EXPECT_EQ(FP(qnan).nan_payload(), FP64::canon);
+    EXPECT_EQ(FP(qnan + 1.0).nan_payload(), FP64::canon);
+    EXPECT_EQ(FP(inf - inf).nan_payload(), FP64::canon);
+    EXPECT_EQ(FP(inf * 0.0).nan_payload(), FP64::canon);
+}
+
+TEST(floating_point_utils, float_nan_payload)
+{
+    constexpr auto inf = FP32::Limits::infinity();
+    constexpr auto qnan = FP32::Limits::quiet_NaN();
+
+    EXPECT_EQ(FP(0.0f).nan_payload(), 0);
+    EXPECT_EQ(FP(FP32::Limits::signaling_NaN()).nan_payload(), 0x200000);
```

Comment:
The exact payloads of `Limits::quiet_NaN()` and `Limits::signaling_NaN()` are implementation defined, aren't they? So maybe not worth it to put it into tests? Or maybe add a comment about this.

Reply:
Moved to separate tests.